packages/jest-cli/src/TestNamePatternModeController.js

```diff
@@ -0,0 +1,164 @@
+
+  if (!match) {
+    return chalk.dim(testName);
+  }
+
+  // $FlowFixMe
+  const start = match.index;
+  const end = start + match[0].length;
+
+  return colorize(testName, Math.max(start, 0), Math.max(end, end));
+};
+
+const testNamePatternModeController = (
+  config: Config,
+  pipe: stream$Writable | tty$WriteStream,
+  promptController: PromptController,
+) => {
+  class TestNamePatternModeController {
+    cachedTestNames: Array<string>;
+
+    run(
+      onSuccess: Function,
+      onCancel: Function,
+    ) {
+      pipe.write(ansiEscapes.cursorHide);
+      pipe.write(ansiEscapes.clearScreen);
+      pipe.write(usage());
+      pipe.write(ansiEscapes.cursorShow);
+
+      promptController.prompt(
+        this.onChange.bind(this),
+        onSuccess,
+        onCancel,
+      );
+    }
+
+    onChange(
+      pattern: string
+    ) {
+      pipe.write(ansiEscapes.eraseLine);
+      pipe.write(ansiEscapes.cursorLeft);
+      this.printTypeahead(pattern);
+    }
+
+    printTypeahead(
+      pattern: string,
+      max: number = 10
+    ) {
+      const matchedTestNames = this.getMatchedTestNames(pattern);
+
+      const total = matchedTestNames.length;
+      const results = matchedTestNames.slice(0, max);
+      const inputText = `${chalk.dim(' pattern \u203A')} ${pattern}`;
+
+      pipe.write(ansiEscapes.eraseDown);
+      pipe.write(inputText);
+      pipe.write(ansiEscapes.cursorSavePosition);
+
+      if (pattern) {
+        if (total) {
+          pipe.write(`\n\n Pattern matches ${total} ${pluralizeTest(total)}.`);
+        } else {
+          pipe.write(`\n\n Pattern matches no tests.`);
+        }
+
+        results
+          .map(
+            testName => formatTestNameByPattern(testName, pattern)
+          )
+          .forEach(
+            testName => pipe.write(`\n  ${chalk.dim('\u203A')} ${testName}`)
+          );
+
+        if (total > max) {
+          const more = total - max;
+          pipe.write(
+            // eslint-disable-next-line max-len
+            `\n  ${chalk.dim(`\u203A and ${more} more ${pluralizeTest(more)}`)}`,
+          );
+        }
+      } else {
+        // eslint-disable-next-line max-len
+        pipe.write(`\n\n ${chalk.italic.yellow('Start typing to filter by a testname regex pattern.')}`);
+      }
+
+      pipe.write(ansiEscapes.cursorTo(stringLength(inputText), usageRows - 1));
+      pipe.write(ansiEscapes.cursorRestorePosition);
+    }
+
+    getMatchedTestNames(pattern: string) {
+      let regex;
+
+      try {
+        regex = new RegExp(pattern, 'i');
+      } catch (e) {
+        return [];
+      }
+
+      return this.cachedTestNames.filter(testName => regex.test(testName));
+    }
+
+    // $FlowFixMe
+    updateCachedTestNames(testResults = []) {
+      this.cachedTestNames = testResults.reduce((testNames, item) =>
+        [
+          ...testNames,
+          ...item.testResults.map(({title: name}) => name),
+        ]
+      , []);
+    }
```

Comment:
Can we make this use a bit less memory?

this._testNames = [];
testResults.forEach(
  testNames => this._testNames.push(...item.testResults.map(({title: name}) => name))
);

that should be more efficient.